### PR TITLE
DAOS-2262 mgmt: Pass in ACL to create pool

### PR DESCRIPTION
**Pull request by kjacque**
- Add support for ACL file parameter in daos_shell pool
  create command. The ACL file should have a single ACE
  on each line, in the short ACL string format.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Comment by mjmac** [comment]
Why bother with all of the mock file operations? It's trivial to create a temp dir for testing and then run your test file operations on files created there. That way you're testing real code paths vs mock code, and it also reduces the amount of non-stdlib code that needs to be maintained.

**Comment by kjacque** [comment]
> Why bother with all of the mock file operations? It's trivial to create a temp dir for testing and then run your test file operations on files created there. That way you're testing real code paths vs mock code, and it also reduces the amount of non-stdlib code that needs to be maintained.

I wanted make it easy to exercise the corner cases. What kind of situation causes the bufio Scanner to error after reading some bytes? I'm not sure. But that is a real error we should check for, and I can mock it. I'm also not crazy about unit tests changing the state of the real system unless it is absolutely necessary.

If you're firmly against the mocks, I'll change it. But in general I think basing stdlib usage around (thin) interfaces and designing for dependency injection is good practice.

**Comment by mjmac** [comment]
> I wanted make it easy to exercise the corner cases. What kind of situation causes the bufio Scanner to error after reading some bytes? I'm not sure. But that is a real error we should check for, and I can mock it. I'm also not crazy about unit tests changing the state of the real system unless it is absolutely necessary.
> 
> If you're firmly against the mocks, I'll change it. But in general I think basing stdlib usage around (thin) interfaces and designing for dependency injection is good practice.

Ok, fair enough. Another approach would be separate the code for doing file management and file parsing into two different areas of concern. Your code for parsing would just accept an io.Reader which would either be the opened file in the normal case or your test thing which implements io.Reader but can simulate corner cases. With that approach the amount of custom code we have to maintain is smaller, and you're not inventing (and propagating throughout the code) new interfaces to enable testing mocks.

**Comment by kjacque** [comment]
> Another approach would be separate the code for doing file management and file parsing into two different areas of concern. Your code for parsing would just accept an io.Reader which would either be the opened file in the normal case or your test thing which implements io.Reader but can simulate corner cases.

This works for me. I'll refactor with this approach. Thanks.

> If your code was doing something special to recover from errors I'd say maybe it's warranted, but it seems like you're just looking for an error and saying "yep, there was an error".

Yeah, we just abort and return the error. But that's how we handle most errors from outside sources, which I don't think it's silly to test when it's plausible to do so.

File: src/mgmt/srv.c [hunk 1]
```diff
@@ -410,15 +411,61 @@ process_join_request(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
...
+	for (i = 0; i < ace_nr; i++) {
+		rc = daos_ace_from_str(ace_strs[i], &(tmp_aces[i]));
+		if (rc != 0) {
+			D_ERROR("Failed to convert string '%s' to ACE, "
+				"err=%d\n", ace_strs[i], rc);
+			D_GOTO(out, rc);
+		}
+	}
+
+	tmp_acl = daos_acl_create(tmp_aces, ace_nr);
```
Comment (tanabarr) [inline src/mgmt/srv.c]: does tmp_acl need to be freed after creating on failure?

File: src/control/cmd/dmg/acl_test.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	if result != nil {
+		t.Error("Expected no result, got something back")
+	}
+
+	if err == nil {
+		t.Fatal("Expected an error, got nil")
+	}
+}
+
+func createTestFile(t *testing.T, filePath string, content string) {
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/cmd/dmg/acl_test.go]: Agree, to a point. I think in some cases a bit of duplication leads to improved clarity, as the savings of having a common implementation is offset by the burden of needing to go look at how it's implemented in order to figure out how it works. Particularly for test code like this, where most of the time it's being read after a test failure in order to figure out how the test works and what went wrong.

File: src/mgmt/srv.c [hunk 3]
```diff
@@ -410,15 +411,61 @@ process_join_request(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
...
+		D_ERROR("Failed to allocate ACL\n");
+		D_GOTO(out, rc = -DER_NOMEM);
+	}
+
+	*acl = tmp_acl;
+	rc = 0;
+
+out:
+	for (i = 0; i < ace_nr; i++)
+		daos_ace_free(tmp_aces[i]);
```
Comment (ashleypittman) [inline src/mgmt/srv.c]: If acr_nr > 2 and a intermediate daos_ace_from_str() call fails then this looks like it'll call daos_acr_free() with uninitialised data for larger values of i.

File: src/mgmt/srv.c (hunk 1 above)
Comment (tanabarr) [reply src/mgmt/srv.c]: nvm , no failure paths after

File: src/mgmt/srv.c [hunk 4]
```diff
@@ -410,15 +411,61 @@ process_join_request(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 }
 
 static int
-create_ownership_props(daos_prop_t **out_prop, char *owner, char *owner_grp)
+ace_strs_to_acl(char **ace_strs, size_t ace_nr, struct daos_acl **acl)
 {
-	char	       *out_owner = NULL;
-	char	       *out_owner_grp = NULL;
-	daos_prop_t    *new_prop = NULL;
+	struct daos_ace	*tmp_aces[ace_nr];
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/mgmt/srv.c]: Moving allocation to the heap

File: src/mgmt/srv.c (hunk 3 above)
Comment (kjacque) [reply src/mgmt/srv.c]: Changing to allocate on the heap with D_ALLOC_ARRAY to avoid VLAs on the stack. This should also initialize all the array pointers to NULL.

File: src/mgmt/srv.c [hunk 5]
```diff
@@ -410,15 +411,64 @@ process_join_request(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
...
 
 static int
-create_ownership_props(daos_prop_t **out_prop, char *owner, char *owner_grp)
+ace_strs_to_acl(char **ace_strs, size_t ace_nr, struct daos_acl **acl)
 {
-	char	       *out_owner = NULL;
-	char	       *out_owner_grp = NULL;
-	daos_prop_t    *new_prop = NULL;
+	struct daos_ace	**tmp_aces;
+	struct daos_acl	*tmp_acl = NULL;
```
Comment (ashleypittman) [inline src/mgmt/srv.c]: I don't think this needs to be initialised now.

File: src/mgmt/pool.pb-c.h [hunk 6]
```diff
@@ -54,10 +54,15 @@ struct  _Mgmt__CreatePoolReq
    * DAOS system identifier
    */
   char *sys;
+  /*
+   * Access Control Entries in short string format
+   */
+  size_t n_acl;
```
Comment (ashleypittman) [inline src/mgmt/pool.pb-c.h]: Why size_t?  This would be better as an unsigned int wouldn't it?

Comment (kjacque) [reply src/mgmt/pool.pb-c.h]: Can't control this one, unfortunately. It's generated code.

File: src/mgmt/srv.c (hunk 5 above)
Comment (kjacque) [reply src/mgmt/srv.c]: Fixed

**Comment by jolivier23** [comment]
Looks like the branch has conflicts that need to be resolved before it can be landed
